samza-core/src/main/java/org/apache/samza/zk/ZkDistributedReadWriteLock.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.zk;
+
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.I0Itec.zkclient.IZkStateListener;
+import org.apache.samza.SamzaException;
+import org.apache.samza.coordinator.DistributedReadWriteLock;
+import org.apache.zookeeper.Watcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Distributed lock primitive for Zookeeper.
+ */
+public class ZkDistributedReadWriteLock implements DistributedReadWriteLock {
+
+  public static final Logger LOG = LoggerFactory.getLogger(ZkDistributedReadWriteLock.class);
+  private static final String PARTICIPANTS_PATH = "participants";
+  private static final String PROCESSORS_PATH = "processors";
+  private final ZkUtils zkUtils;
+  private final String lockPath;
+  private final String particpantsPath;
+  private final String processorsPath;
+  private final String participantId;
+  private final ZkKeyBuilder keyBuilder;
+  private final Random random = new Random();
+  private String activeParticipantPath = null;
+  private String activeProcessorPath = null;
+  private Object mutex;
```

Comment:
final

Reply:
made mutex final. 
the paths (activeParticipantPath and activeProcessorPath) can not be inited in the constructor and are set to null when the corresponding ephemeral nodes are deleted. Thus these can not be made final.